### DISCUSSION TITLE
링버퍼 관련 질문

### DISCUSSION THREAD

**Discussion by ghost** [luftaquila/monolith#20] [Q&A] [answered]
안녕하세요, 병준님!
국민대 KUST 팀입니다.

그동안 데이터로거에 대해서 공부를 해보았는데, 다른 코드랑은 다르게 Ringbuffer라는 녀석이 눈에 띄더라구요.
링버퍼는 찾아보니 많이들 usart 기능을 사용하다가 임베디드 시스템에서 좀 더 동적인 메모리 할당을 위해 사용한다고 들었는데,
생소한 개념이라 열심히 공부하고 있습니다!

구글링해보니 링버퍼를 이용한 코드가 많지는 않던데 정확히 어떤 필요성에 의해서 링버퍼를 사용하셨는지 궁금합니다!

**Comment by luftaquila** [comment, marked as answer]
ring buffer의 다른 이름은 circular queue입니다. 모노리스에서는 로그를 SD카드에 적기 위한 쓰기 버퍼나, UART와 I2C로 보내기 위한 전송 버퍼에 쓰고 있는데요. 기본적으로 로그가 발생하는 속도보다 sdcard write나 uart/i2c transmit 주기가 길기 때문에 그 사이에 쌓이는 로그를 담아두기 위한 큐 입니다.

어쨌거나 버퍼는 용량이 정해져 있어서 꽉 차면 처리를 해줘야 하는데요, 원형 큐를 쓰면 자연스럽게 가장 오래된 데이터부터 덮어쓰기 때문에 사용하고 있습니다. 잡다한 구현이 번거로워서 Ring-Buffer라는 구현체를 가져다 사용하고 있고요.

더 자세한 내용은 자료구조 관련 블로그 등에 정보가 많으니 찾아보시면 될 듯합니다. OS가 올라가는 임베디드 시스템에서는 OS가 메모리 관리를 해주니 동적 할당이라는 개념이 존재하지만, 그냥 펌웨어 수준에서는 물리적인 메모리 주소를 그대로 바라보기 때문에 동적 할당은 모노리스에서는 무관한 주제라고 생각하시면 될 듯합니다 🙂
